src/library/scala/collection/TraversableLike.scala

```diff
@@ -451,17 +452,19 @@ trait TraversableLike[+A, +Repr] extends Any
   }
 
   def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
-    val m = mutable.Map.empty[K, Builder[A, Repr]]
+    val m = new HashMapBuilder[K, Builder[A, Repr]]
     for (elem <- this) {
       val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newBuilder)
+      val bldr = m.getOrElse(key, null) match {
+        case null =>
+          val b = newBuilder
+          m.+=((key, b))
```

Comment:
It is a bit more work, and something of an orthoganal improvement, to avoid the intermediate `Tuple2` here. We need to thread the key, value, and an optional tuple through the creation methods in `HashMap` (ie `addOne(k: K, v: V, kvOrNull: (K, V)`).

Sketch: https://github.com/scala/scala/pull/8935/files

This PR stands on its own though.